File: year_2023/day_04/day04.py
import math
from collections import deque

import regex as re

from tools import loader, parsers


def lottery(data: list[str]) -> tuple[int, int]:
    """
    >>> print(lottery(parsers.lines('test.txt')))
    (13, 30)"""
    tickets = {}
    for i, line in enumerate(data, start=1):
        winning, ticket = line.split(':')[1].split('|')
        winning = set(map(int, re.findall(r'\d+', winning)))
        ticket = set(map(int, re.findall(r'\d+', ticket)))
        tickets[i] = (winning, ticket)

    part1 = part2 = 0
    for winning, ticket in tickets.values():
        if matches := winning.intersection(ticket):
            part1 += int(math.pow(2, len(matches) - 1))

    queue = deque([*tickets])
    seen = {}
    while queue:
        index = queue.popleft()
        part2 += 1
        if index not in seen:
            winning, ticket = tickets[index]
            matches = len(winning.intersection(ticket))
            new_tickets = [index + i + 1 for i in range(matches)]
            seen[index] = new_tickets
        queue.extend(seen[index])

    return part1, part2


print(lottery(parsers.lines(loader.get())))  # 32001, 5037841

File: year_2023/day_04/day04.go
package main

import (
	"aoc/tools"
	"fmt"
	"math"
	"strings"
)

type card struct {
	winning []int
	ticket  []int
}

func (c card) intersect() int {
	result := 0
	winning := map[int]bool{}

	for _, num := range c.winning {
		winning[num] = true
	}

	for _, val := range c.ticket {
		if winning[val] {
			result++
		}
	}

	return result
}

func solvePart1(tickets map[int]card) int {
	var result int

	for _, ticket := range tickets {
		if matches := ticket.intersect(); matches > 0 {
			result += int(math.Pow(2, float64(matches-1)))
		}
	}

	return result
}

func solvePart2(tickets map[int]card) int {
	result := 0
	matches := map[int][]int{}
	queue := []int{}

	for index := range tickets {
		queue = append(queue, index)
	}

	for len(queue) > 0 {
		current := queue[0]
		result++
		queue = queue[1:]
		wins, ok := matches[current]

		if !ok {
			for i := current; i < current+tickets[current-1].intersect(); i++ {
				wins = append(wins, i+1)
			}

			matches[current] = wins
		}

		queue = append(queue, wins...)
	}

	return result
}

func solve(input string) (int, int) {
	tickets := map[int]card{}

	for i, line := range tools.ReadLines(input) {
		split := strings.FieldsFunc(line, func(r rune) bool { return r == ':' || r == '|' })
		tickets[i+1] = card{
			winning: tools.StrToInt(strings.Fields(split[1])),
			ticket:  tools.StrToInt(strings.Fields(split[2])),
		}
	}

	return solvePart1(tickets), solvePart2(tickets)
}

func main() {
	fmt.Println(solve("test.txt"))
	fmt.Println(solve(tools.GetData(2023, 4)))
}
